wip: API for projects and boards

dineshsalunke opened the pull request:
Fixes #14299 

This is my first jab at anything in go lang, so please bear with me if you notice some mistakes for time being.

- [x] Swagger documentation for all the required endpoints.
- [x] Integration Test suites for all the endpoints.
- [x] Implementation's

6543 commented on integrations/api_repo_project_test.go:
```diff
@@ -0,0 +1,13 @@
+)
+
+func TestAPIRepoProjects(t *testing.T) {
+	defer prepareTestEnv(t)()
```
an open TODO ...

6543 commented on routers/api/v1/swagger/repo.go:
```diff
@@ -351,3 +351,27 @@ type swaggerRepoCollaboratorPermission struct {
 	Body api.RepoCollaboratorPermission `json:"body"`
 }
+
+// swagger:response Project
+type swaggerProject struct {
+	// in:body
+	Body api.Project `json:"body"`
+}
+
+// swagger:response ProjectList
+type swaggerProjectList struct {
+	// in:body
+	Body []api.Project `json:"body"`
+}
+
+// swagger:response ProjectBoard
+type swaggerProjectBoard struct {
+	// in:body
+	Body api.ProjectBoard `json:"body"`
+}
+
+// swagger:response ProjectBoardList
```
```suggestion
// Project
// swagger:response Project
type swaggerProject struct {
	// in:body
	Body api.Project `json:"body"`
}

// ProjectList
// swagger:response ProjectList
type swaggerProjectList struct {
	// in:body
	Body []api.Project `json:"body"`
}

// ProjectBoard
// swagger:response ProjectBoard
type swaggerProjectBoard struct {
	// in:body
	Body api.ProjectBoard `json:"body"`
}

// ProjectBoardList
// swagger:response ProjectBoardList
```
make it similar to others ...

6543 commented:
> I will work on the the test's and implementations once the API definition is ok, can anyone confirm if the api def is ok

@dineshsalunke just have a look at https://github.com/go-gitea/gitea/blob/main/CONTRIBUTING.md#api-v1

- List endpoint do miss page & limit support

the endpoints do look like they are similar to https://docs.github.com/en/rest/projects/projects#list-repository-projects - so that's ok

6543 commented on integrations/api_repo_project_test.go:
```diff
@@ -0,0 +1,13 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
```
use the same header on all new files:
```go
// Copyright 2022 The Gitea Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.
```

6543 commented on models/project/board.go:
```diff
@@ -224,6 +227,29 @@ func GetBoards(ctx context.Context, projectID int64) (BoardList, error) {
 
+func GetBoardsAndCount(ctx context.Context, projectID int64) (BoardList, int64, error) {
+	engine := db.GetEngine(ctx)
+	boards := make([]*Board, 0, setting.UI.IssuePagingNum)
```
```suggestion
	boards := make([]*Board, 0, 10)
```
we use some number between 5-10 depending on what we thing is most common
if you like to introduce pagination you should do so by adding an option struct

6543 replied:
we migt should refactor that and use a const ...

6543 commented on models/project/board.go:
```diff
@@ -224,6 +227,29 @@ func GetBoards(ctx context.Context, projectID int64) (BoardList, error) {
+		return nil, 0, err
+	}
+
+	count, err := engine.Count(new(Board))
```
```suggestion
	count, err := engine.FindAndCount(&boards)
```

6543 commented on models/project/board.go:
```diff
@@ -224,6 +227,29 @@ func GetBoards(ctx context.Context, projectID int64) (BoardList, error) {
+
+	count, err := engine.Count(new(Board))
+	if err != nil {
+		return nil, 0, fmt.Errorf("count: %v", err)
```
```suggestion
		return nil, 0, err
```

6543 commented on models/project/board.go:
```diff
@@ -224,6 +227,29 @@ func GetBoards(ctx context.Context, projectID int64) (BoardList, error) {
+	if err != nil {
+		return nil, 0, fmt.Errorf("count: %v", err)
+	}
+
+	if err = engine.Find(&boards); err != nil {
+		return nil, 0, fmt.Errorf("internal server error: %v", err)
+	}
+
```
```suggestion
```

6543 commented on models/project/board.go:
```diff
@@ -224,6 +227,29 @@ func GetBoards(ctx context.Context, projectID int64) (BoardList, error) {
+	engine := db.GetEngine(ctx)
+	boards := make([]*Board, 0, setting.UI.IssuePagingNum)
+
+	engine.Where("project_id=? AND `default`=?", projectID, false).OrderBy("Sorting")
```
```suggestion
	engine.Where("project_id=?", projectID).OrderBy("Sorting")
```
use opt struct as mentioned for filters ...

6543 commented on models/project/board.go:
```diff
@@ -224,6 +227,24 @@ func GetBoards(ctx context.Context, projectID int64) (BoardList, error) {
+	engine := db.GetEngine(ctx)
+	boards := make([]*Board, 0, 10)
+
+	engine.Where("project_id=?", projectID).OrderBy("Sorting")
```
```suggestion
	engine.Where("project_id=? AND `default`=?", projectID, false).OrderBy("Sorting")
```

6543 commented on models/project/board.go:
```diff
@@ -224,6 +227,24 @@ func GetBoards(ctx context.Context, projectID int64) (BoardList, error) {
+	if err != nil {
+		return nil, 0, err
+	}
+	return append([]*Board{defaultB}, boards...), count, nil
```
```suggestion
	return append([]*Board{defaultB}, boards...), count+1, nil
```

6543 commented on routers/api/v1/repo/board.go:
```diff
@@ -0,0 +1,227 @@
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+
+	board, err := project_model.GetBoard(ctx, ctx.ParamsInt64(":id"))
```
check if board.RepoID match ctx.Repo

6543 replied:
-> permission check

6543 commented:
@dineshsalunke sorry for the delay :)

lunny commented on routers/api/v1/api.go:
```diff
@@ -938,6 +938,11 @@ func Routes(ctx gocontext.Context) *web.Route {
+				m.Group("/projects", func() {
+					m.Combo("").
+						Get(reqToken(), repo.ListRepositoryProjects).
+						Post(reqToken(), mustNotBeArchived, bind(api.NewProjectPayload{}), repo.CreateRepositoryProject)
```
Permissions check is necessary.

lunny commented on routers/api/v1/api.go:
```diff
@@ -1177,6 +1182,27 @@ func Routes(ctx gocontext.Context) *web.Route {
+					Get(reqToken(), repo.GetProjectBoard).
+					Patch(reqToken(), bind(api.UpdateProjectBoardPayload{}), repo.UpdateProjectBoard).
+					Delete(reqToken(), repo.DeleteProjectBoard)
+			})
```
Permissions check is necessary.

lunny commented on models/project/project.go:
```diff
@@ -339,6 +346,63 @@ func DeleteProjectByIDCtx(ctx context.Context, id int64) error {
+}
+
+// LoadAttributes load repos and creators of projects.
+func (pl List) LoadAttributes(ctx context.Context) (err error) {
```
Moving `List` to a separate file is better.

tobiasBora commented:
Awesome! Do you think it would be possible to add a way to assign a project to a board directly at its creation? It would make scripting easier.

f403 commented:
What is the plan for this PR? This is the main feature I had been waiting in 1.18

dineshsalunke commented:
@f403 sorry I didn't get time to finish this up been busy with a lot of things, with the baby coming in etc..

I will finish this and push it asap

f403 commented:
@dineshsalunke , no problem, thank you.

>  with the baby coming in etc

Congratulations! This is much more important

lunny commented on models/project/project_list.go:
```diff
@@ -0,0 +1,47 @@
+package project
```
copyright head is necessary.

dineshsalunke replied:
oops sorry my bad

dineshsalunke commented:
> Awesome! Do you think it would be possible to add a way to assign a project to a board directly at its creation? (similarly to how label can be added at the creation of an issue) It would make scripting easier.

i dont think that should be difficult let me check on that and get back

lunny commented on routers/api/v1/repo/board.go:
```diff
@@ -0,0 +1,232 @@
+// Copyright 2016 The Gogs Authors. All rights reserved.
+// Copyright 2018 The Gitea Authors. All rights reserved.
```
Since this is a new file, it just needs Gitea Authors copyright and the year should be 2022

lunny commented on routers/api/v1/repo/project.go:
```diff
@@ -0,0 +1,241 @@
+// Copyright 2016 The Gogs Authors. All rights reserved.
+// Copyright 2018 The Gitea Authors. All rights reserved.
```
As above

dineshsalunke replied:
then I think I might have to update it at few more places

dineshsalunke replied:
what about this

![image](https://user-images.githubusercontent.com/52815633/209367838-25ab5750-29c3-4165-bf98-be3ab6261810.png)

I noticed this at many places I copied from it from there

lunny commented on routers/api/v1/repo/board.go:
```diff
@@ -0,0 +1,231 @@
+// Copyright 2022 The Gogs Authors. All rights reserved.
```
Please remove this line.

techknowlogick replied:
```suggestion
```

lunny commented on routers/api/v1/repo/board.go:
```diff
@@ -0,0 +1,231 @@
+		return
+	}
+
+	if board.Title != form.Title {
```
Looks like the check is unnecessary, we can just assign the value.

lunny replied:
And check the board's repository id.

dineshsalunke replied:
what error to return if the repo ids don't match. Unauthorized of BadRequest ?

lunny replied:
If that, `reqRepoReader(unit.TypeProjects),` will not get the right repository and will not check the right actually. We need check the permission in the function.

lunny commented on routers/api/v1/repo/board.go:
```diff
@@ -0,0 +1,231 @@
+	//     "$ref": "#/responses/forbidden"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+	if err := project_model.DeleteBoardByID(ctx.ParamsInt64(":id")); err != nil {
```
You should also check the board's repository id is equal to context repository id.

dineshsalunke replied:
I don't think we have access to the repository on these routes since these routes are not repo or project dependent

techknowlogick replied:
@lunny There is the permission check above, to confirm user can write to projects, and it is based on the repo linked to the project.

lunny commented on routers/api/v1/repo/board.go:
```diff
@@ -0,0 +1,231 @@
+	//     "$ref": "#/responses/forbidden"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+	boards, count, err := project_model.GetBoardsAndCount(ctx, ctx.ParamsInt64(":id"))
```
As above.

dineshsalunke replied:
same as above

lunny commented on routers/api/v1/repo/board.go:
```diff
@@ -0,0 +1,231 @@
+		Default:   form.Default,
+		Sorting:   form.Sorting,
+		Color:     form.Color,
+		ProjectID: ctx.ParamsInt64(":id"),
```
As above.

dineshsalunke replied:
same as above

lunny commented on routers/api/v1/repo/project.go:
```diff
@@ -0,0 +1,240 @@
+	//     "$ref": "#/responses/forbidden"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+	project, err := project_model.GetProjectByID(ctx, ctx.ParamsInt64(":id"))
```
As above.

dineshsalunke replied:
same as above

dineshsalunke replied:
what error should we return in this case, Unauthorized, BadRequest or NotFound ?

lunny commented on routers/api/v1/repo/project.go:
```diff
@@ -0,0 +1,240 @@
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+	form := web.GetForm(ctx).(*api.UpdateProjectPayload)
+	project, err := project_model.GetProjectByID(ctx, ctx.ParamsInt64(":id"))
```
As above.

dineshsalunke replied:
same as above

dineshsalunke replied:
what error should we return in this case, Unauthorized, BadRequest or NotFound ?

lunny commented on routers/api/v1/repo/project.go:
```diff
@@ -0,0 +1,240 @@
+	//     "$ref": "#/responses/forbidden"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+	err := project_model.DeleteProjectByID(ctx, ctx.ParamsInt64(":id"))
```
As above.

dineshsalunke replied:
what error should we return in this case, Unauthorized, BadRequest or NotFound ?

techknowlogick commented on modules/structs/project.go:
```diff
@@ -0,0 +1,69 @@
+// Copyright 2014 The Gogs Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
```
```suggestion
// Copyright 2023 The Gitea Authors. All rights reserved.
// SPDX-License-Identifier: MIT
```

techknowlogick commented on routers/api/v1/repo/project.go:
```diff
@@ -0,0 +1,261 @@
+// Copyright 2022 The Gogs Authors. All rights reserved.
```
```suggestion
```

techknowlogick commented on routers/api/v1/api.go:
```diff
@@ -985,6 +985,11 @@ func Routes(ctx gocontext.Context) *web.Route {
 				}, mustEnableIssuesOrPulls)
+				m.Group("/projects", func() {
+					m.Combo("").
+						Get(reqToken(auth_model.AccessTokenScopeRepo), repo.ListRepositoryProjects).
```
I've used Repo Scope for project boards, but this probably needs to be reviewed as to which scope is most appropriate.

f403 replied:
Are [these](/go-gitea/gitea/blob/main/docs/content/doc/developers/oauth2-provider.en-us.md) all possible Scopes? There is nothing else except `AccessTokenScopeRepo` to fit then

f403 commented:
Do I understand correctly, this won't be merged in 1.19.0?

lunny commented:
> Do I understand correctly, this PR won't be merged in 1.19.0?

Now it's in milestone 1.20

lunny commented on routers/api/v1/repo/project.go:
```diff
@@ -0,0 +1,260 @@
+	//     "$ref": "#/responses/forbidden"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+	fmt.Print(ctx.FormOptionalBool("closed"))
```
Please remove the trace line.

6543 commented on models/project/project.go:
```diff
@@ -85,6 +85,9 @@ func (err ErrProjectBoardNotExist) Unwrap() error {
 	return util.ErrNotExist
 }
 
+// List is a list of projects
```
?

6543 commented on models/project/project_list.go:
```diff
@@ -0,0 +1,49 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
```
```suggestion
// Copyright 2023 The Gitea Authors. All rights reserved.
```

...

6543 commented:
TODOs:
- [ ] Fix Tests
- [ ] Optional: handle org/individual projects
- [ ] make board api similar to github (/projects/{projectID}/columns) 
- [ ] merge "board" categorty into "project" category in swagger docs
- [ ] /projects/columns/{boardID}/cards for actuall issues/pulls list to return

GamerGirlandCo commented:
i've taken it upon myself to bring this pr up to date with the base branch and fix some panics i encountered. attached is a patch file you can apply @dineshsalunke 

[gitea-projects-fixes.patch](https://github.com/go-gitea/gitea/files/12270442/gitea-projects-fixes.patch)

dineshsalunke commented:
first of all sorry for the delay in response, a lot happened in past six months and couldn't find time.

@tgy that is pretty much the exact case for which I started this PR, this could be really useful in our startup.
@GamerGirlandCo thanks for the patch I will check on it and update the PR accordingly

I will try to stay on top of this and get it completed soon

tgy commented:
i ended up moving from the Project functionality to simply use the Issues to manage tasks

accforgithubtest commented:
would love to see this released !

dineshsalunke commented:
@6543 I have moved the methods into its own packages and also added new endpoints can you do a quick check on it and its ok then I will move forward with updating all the routes accordingly

lng2020 commented:
Any progress? I can help you with writing tests or something. It's been a really old PR.

dineshsalunke commented:
@lng2020 Sorry for the delay, we are closing on private alpha on our project so were busy with that.

> I can help you with writing tests or something. It's been a really old PR.

I would be more than happy feel free to add things as you feel.

> Is it okay to close this one and reopen the same one? The commit history is too long and mixed with some other information.

honestly, I was thinking of the same thing its just that I didn't wanted history to repeat itself :D so I am thinking of working on finishing up the major stuff in here itself and then once we are in good position then we can close this and raise a new PR.

dineshsalunke commented:
@lng2020 If you don't mind me asking, what do you plan on using these changes for ? I mean do you wanna integrate something with gitea in terms of projects and issues ?

lng2020 commented:
> What do you plan on using these changes for?
Nothing particular. I'm just thrilled to see Gitea become more and more complete.
